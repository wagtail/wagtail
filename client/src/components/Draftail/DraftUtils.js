
/**
* Returns collection of currently selected blocks.
* See https://github.com/jpuri/draftjs-utils/blob/e81c0ae19c3b0fdef7e0c1b70d924398956be126/js/block.js#L19.
*/
const getSelectedBlocksList = (editorState) => {
  const selectionState = editorState.getSelection();
  const content = editorState.getCurrentContent();
  const startKey = selectionState.getStartKey();
  const endKey = selectionState.getEndKey();
  const blockMap = content.getBlockMap();
  const blocks =  blockMap
    .toSeq()
    .skipUntil((_, k) => k === startKey)
    .takeUntil((_, k) => k === endKey)
    .concat([[endKey, blockMap.get(endKey)]]);
  return blocks.toList();
};

/**
* Returns the currently selected text in the editor.
* See https://github.com/jpuri/draftjs-utils/blob/e81c0ae19c3b0fdef7e0c1b70d924398956be126/js/block.js#L106.
*/
export const getSelectionText = (editorState) => {
  const selection = editorState.getSelection();
  let start = selection.getAnchorOffset();
  let end = selection.getFocusOffset();
  const selectedBlocks = getSelectedBlocksList(editorState);

  if (selection.getIsBackward()) {
    const temp = start;
    start = end;
    end = temp;
  }

  let selectedText = '';
  for (let i = 0; i < selectedBlocks.size; i += 1) {
    const blockStart = i === 0 ? start : 0;
    const blockEnd = i === (selectedBlocks.size - 1) ? end : selectedBlocks.get(i).getText().length;
    selectedText += selectedBlocks.get(i).getText().slice(blockStart, blockEnd);
  }

  return selectedText;
};
