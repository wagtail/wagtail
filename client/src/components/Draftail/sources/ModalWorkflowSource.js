import PropTypes from 'prop-types';
import { Component } from 'react';
import { AtomicBlockUtils, Modifier, RichUtils, EditorState } from 'draft-js';
import { ENTITY_TYPE } from 'draftail';

import { STRINGS } from '../../../config/wagtailConfig';
import { getSelectionText } from '../DraftUtils';

const $ = global.jQuery;

const EMBED = 'EMBED';
const DOCUMENT = 'DOCUMENT';

const MUTABILITY = {};
MUTABILITY[ENTITY_TYPE.LINK] = 'MUTABLE';
MUTABILITY[DOCUMENT] = 'MUTABLE';
MUTABILITY[ENTITY_TYPE.IMAGE] = 'IMMUTABLE';
MUTABILITY[EMBED] = 'IMMUTABLE';

export const getChooserConfig = (entityType, entity, selectedText) => {
  let url;
  let urlParams;

  switch (entityType.type) {
  case ENTITY_TYPE.IMAGE:
    return {
      url: `${global.chooserUrls.imageChooser}?select_format=true`,
      urlParams: {},
      onload: global.IMAGE_CHOOSER_MODAL_ONLOAD_HANDLERS,
    };

  case EMBED:
    return {
      url: global.chooserUrls.embedsChooser,
      urlParams: {},
      onload: global.EMBED_CHOOSER_MODAL_ONLOAD_HANDLERS,
    };

  case ENTITY_TYPE.LINK:
    url = global.chooserUrls.pageChooser;
    urlParams = {
      page_type: 'wagtailcore.page',
      allow_external_link: true,
      allow_email_link: true,
      can_choose_root: 'false',
      link_text: selectedText,
    };

    if (entity) {
      const data = entity.getData();

      if (data.id) {
        url = `${global.chooserUrls.pageChooser}${data.parentId}/`;
      } else if (data.url.startsWith('mailto:')) {
        url = global.chooserUrls.emailLinkChooser;
        urlParams.link_url = data.url.replace('mailto:', '');
      } else {
        url = global.chooserUrls.externalLinkChooser;
        urlParams.link_url = data.url;
      }
    }

    return {
      url,
      urlParams,
      onload: global.PAGE_CHOOSER_MODAL_ONLOAD_HANDLERS,
    };

  case DOCUMENT:
    return {
      url: global.chooserUrls.documentChooser,
      urlParams: {},
      onload: global.DOCUMENT_CHOOSER_MODAL_ONLOAD_HANDLERS,
    };

  default:
    return {
      url: null,
      urlParams: {},
      onload: {},
    };
  }
};

export const filterEntityData = (entityType, data) => {
  switch (entityType.type) {
  case ENTITY_TYPE.IMAGE:
    return {
      id: data.id,
      src: data.preview.url,
      alt: data.alt,
      format: data.format,
    };
  case EMBED:
    return {
      embedType: data.embedType,
      url: data.url,
      providerName: data.providerName,
      authorName: data.authorName,
      thumbnail: data.thumbnail,
      title: data.title,
    };
  case ENTITY_TYPE.LINK:
    if (data.id) {
      return {
        url: data.url,
        id: data.id,
        parentId: data.parentId,
      };
    }

    return {
      url: data.url,
    };
  case DOCUMENT:
    return {
      url: data.url,
      filename: data.filename,
      id: data.id,
    };
  default:
    return {};
  }
};

/**
 * Interfaces with Wagtail's ModalWorkflow to open the chooser,
 * and create new content in Draft.js based on the data.
 */
class ModalWorkflowSource extends Component {
  constructor(props) {
    super(props);

    this.onChosen = this.onChosen.bind(this);
    this.onClose = this.onClose.bind(this);
  }

  componentDidMount() {
    const { onClose, entityType, entity, editorState } = this.props;
    const selectedText = getSelectionText(editorState);
    const { url, urlParams, onload } = getChooserConfig(entityType, entity, selectedText);

    $(document.body).on('hidden.bs.modal', this.onClose);

    // eslint-disable-next-line new-cap
    this.workflow = global.ModalWorkflow({
      url,
      urlParams,
      onload,
      responses: {
        imageChosen: this.onChosen,
        // Discard the first parameter (HTML) to only transmit the data.
        embedChosen: (_, data) => this.onChosen(data),
        documentChosen: this.onChosen,
        pageChosen: this.onChosen,
      },
      onError: () => {
        // eslint-disable-next-line no-alert
        window.alert(STRINGS.SERVER_ERROR);
        onClose();
      },
    });
  }

  componentWillUnmount() {
    this.workflow = null;

    $(document.body).off('hidden.bs.modal', this.onClose);
  }

  onChosen(data) {
    const { editorState, entityType, onComplete } = this.props;
    const content = editorState.getCurrentContent();
    const selection = editorState.getSelection();

    const entityData = filterEntityData(entityType, data);
    const mutability = MUTABILITY[entityType.type];
    const contentWithEntity = content.createEntity(entityType.type, mutability, entityData);
    const entityKey = contentWithEntity.getLastCreatedEntityKey();

    let nextState;

    if (entityType.block) {
      // Only supports adding entities at the moment, not editing existing ones.
      // See https://github.com/springload/draftail/blob/cdc8988fe2e3ac32374317f535a5338ab97e8637/examples/sources/ImageSource.js#L44-L62.
      // See https://github.com/springload/draftail/blob/cdc8988fe2e3ac32374317f535a5338ab97e8637/examples/sources/EmbedSource.js#L64-L91
      nextState = AtomicBlockUtils.insertAtomicBlock(editorState, entityKey, ' ');
    } else {
      // Replace text if the chooser demands it, or if there is no selected text in the first place.
      const shouldReplaceText = data.prefer_this_title_as_link_text || selection.isCollapsed();

      if (shouldReplaceText) {
        // If there is a title attribute, use it. Otherwise we inject the URL.
        const newText = data.title || data.url;
        const newContent = Modifier.replaceText(content, selection, newText, null, entityKey);
        nextState = EditorState.push(editorState, newContent, 'insert-characters');
      } else {
        nextState = RichUtils.toggleLink(editorState, selection, entityKey);
      }
    }

    // IE11 crashes when rendering the new entity in contenteditable if the modal is still open.
    // Other browsers do not mind. This is probably a focus management problem.
    // From the user's perspective, this is all happening too fast to notice either way.
    this.workflow.close();

    onComplete(nextState);
  }

  onClose(e) {
    const { onClose } = this.props;
    e.preventDefault();

    onClose();
  }

  render() {
    return null;
  }
}

ModalWorkflowSource.propTypes = {
  editorState: PropTypes.object.isRequired,
  entityType: PropTypes.object.isRequired,
  entity: PropTypes.object,
  onComplete: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
};

ModalWorkflowSource.defaultProps = {
  entity: null,
};

export default ModalWorkflowSource;
