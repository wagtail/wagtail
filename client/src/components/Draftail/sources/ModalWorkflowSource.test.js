import React from 'react';
import { shallow } from 'enzyme';

import ModalWorkflowSource, { getChooserConfig, filterEntityData } from './ModalWorkflowSource';
import * as DraftUtils from '../DraftUtils';
import { EditorState, convertFromRaw, AtomicBlockUtils, RichUtils, Modifier } from 'draft-js';

global.ModalWorkflow = () => {};

describe('ModalWorkflowSource', () => {
  beforeEach(() => {
    jest.spyOn(global, 'ModalWorkflow');
    jest.spyOn(DraftUtils, 'getSelectionText').mockImplementation(() => '');
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('works', () => {
    expect(shallow((
      <ModalWorkflowSource
        editorState={{}}
        entityType={{}}
        entity={{}}
        onComplete={() => {}}
        onClose={() => {}}
      />
    ))).toMatchSnapshot();
  });

  describe('#getChooserConfig', () => {
    it('IMAGE', () => {
      expect(getChooserConfig({ type: 'IMAGE' }, null, '')).toEqual({
        url: '/admin/images/chooser/?select_format=true',
        urlParams: {},
        onload: global.IMAGE_CHOOSER_MODAL_ONLOAD_HANDLERS,
      });
    });

    it('EMBED', () => {
      expect(getChooserConfig({ type: 'EMBED' }, null, '')).toEqual({
        url: '/admin/embeds/chooser/',
        urlParams: {},
        onload: global.EMBED_CHOOSER_MODAL_ONLOAD_HANDLERS,
      });
    });

    it('DOCUMENT', () => {
      expect(getChooserConfig({ type: 'DOCUMENT' }, null, '')).toEqual({
        url: '/admin/documents/chooser/',
        urlParams: {},
        onload: global.DOCUMENT_CHOOSER_MODAL_ONLOAD_HANDLERS,
      });
    });

    describe('LINK', () => {
      it('no entity', () => {
        expect(getChooserConfig({ type: 'LINK' }, null, '')).toMatchSnapshot();
      });

      it('page', () => {
        expect(getChooserConfig({ type: 'LINK' }, {
          getData: () => ({ id: 1, parentId: 0 })
        }, '')).toMatchSnapshot();
      });

      it('mail', () => {
        expect(getChooserConfig({ type: 'LINK' }, {
          getData: () => ({ url: 'mailto:test@example.com' })
        }, '')).toMatchSnapshot();
      });

      it('external', () => {
        expect(getChooserConfig({ type: 'LINK' }, {
          getData: () => ({ url: 'https://www.example.com/' })
        }, '')).toMatchSnapshot();
      });
    });
  });

  describe('#filterEntityData', () => {
    it('IMAGE', () => {
      expect(filterEntityData({ type: 'IMAGE' }, {
        id: 53,
        title: 'Test',
        alt: 'Test',
        class: 'richtext-image right',
        edit_link: '/admin/images/53/',
        format: 'right',
        preview: {
          url: '/media/images/test.width-500.jpg',
        }
      })).toMatchSnapshot();
    });

    it('EMBED', () => {
      expect(filterEntityData({ type: 'EMBED' }, {
        authorName: 'Test',
        embedType: 'video',
        providerName: 'YouTube',
        thumbnail: 'https://i.ytimg.com/vi/pSlVtxLOYiM/hqdefault.jpg',
        title: 'Test',
        url: 'https://www.youtube.com/watch?v=pSlVtxLOYiM',
      })).toMatchSnapshot();
    });

    it('DOCUMENT', () => {
      expect(filterEntityData({ type: 'DOCUMENT' }, {
        edit_link: '/admin/documents/edit/1/',
        filename: 'test.pdf',
        id: 1,
        title: 'Test',
        url: '/documents/1/test.pdf',
      })).toMatchSnapshot();
    });

    it('OTHER', () => {
      expect(filterEntityData({ type: 'OTHER' }, {})).toEqual({});
    });

    describe('LINK', () => {
      it('page', () => {
        expect(filterEntityData({ type: 'LINK' }, {
          id: 60,
          parentId: 1,
          url: '/',
          editUrl: '/admin/pages/60/edit/',
          title: 'Welcome to the Wagtail Bakery!',
        })).toMatchSnapshot();
      });

      it('mail', () => {
        expect(filterEntityData({ type: 'LINK' }, {
          prefer_this_title_as_link_text: false,
          title: 'test@example.com',
          url: 'mailto:test@example.com',
        })).toMatchSnapshot();
      });

      it('external', () => {
        expect(filterEntityData({ type: 'LINK' }, {
          prefer_this_title_as_link_text: false,
          title: 'https://www.example.com/',
          url: 'https://www.example.com/',
        })).toMatchSnapshot();
      });
    });
  });

  it('#componentDidMount', () => {
    const wrapper = shallow((
      <ModalWorkflowSource
        editorState={EditorState.createEmpty()}
        entityType={{}}
        entity={{}}
        onComplete={() => {}}
        onClose={() => {}}
      />
    ));

    wrapper.instance().onChosen = jest.fn();

    wrapper.instance().componentDidMount();

    global.ModalWorkflow.mock.calls[0][0].responses.embedChosen('test', {});

    expect(global.ModalWorkflow).toHaveBeenCalled();
    expect(global.jQuery().on).toHaveBeenCalled();
    expect(wrapper.instance().onChosen).toHaveBeenCalled();
  });

  it('#onError', () => {
    window.alert = jest.fn();
    const onClose = jest.fn();

    const wrapper = shallow((
      <ModalWorkflowSource
        editorState={EditorState.createEmpty()}
        entityType={{}}
        entity={{}}
        onComplete={() => {}}
        onClose={onClose}
      />
    ));

    wrapper.instance().componentDidMount();

    global.ModalWorkflow.mock.calls[0][0].onError();

    expect(global.ModalWorkflow).toHaveBeenCalled();
    expect(global.jQuery().on).toHaveBeenCalled();
    expect(window.alert).toHaveBeenCalled();
    expect(onClose).toHaveBeenCalled();
  });

  it('#componentWillUnmount', () => {
    const wrapper = shallow((
      <ModalWorkflowSource
        editorState={EditorState.createEmpty()}
        entityType={{}}
        entity={{}}
        onComplete={() => {}}
        onClose={() => {}}
      />
    ));

    wrapper.instance().componentWillUnmount();

    expect(global.jQuery().off).toHaveBeenCalled();
  });

  describe('#onChosen', () => {
    it('works', () => {
      jest.spyOn(RichUtils, 'toggleLink');

      const onComplete = jest.fn();
      const close = jest.fn();

      let editorState = EditorState.createWithContent(convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: 'a',
            text: 'test',
          }
        ]
      }));
      let selection = editorState.getSelection();
      selection = selection.merge({
        focusOffset: 4,
      });
      editorState = EditorState.acceptSelection(editorState, selection);
      const wrapper = shallow((
        <ModalWorkflowSource
          editorState={editorState}
          entityType={{}}
          entity={{}}
          onComplete={onComplete}
          onClose={() => {}}
        />
      ));

      wrapper.instance().workflow = { close };
      wrapper.instance().onChosen({});

      expect(onComplete).toHaveBeenCalled();
      expect(RichUtils.toggleLink).toHaveBeenCalled();
      expect(close).toHaveBeenCalled();

      RichUtils.toggleLink.mockRestore();
    });

    it('block', () => {
      jest.spyOn(AtomicBlockUtils, 'insertAtomicBlock');

      const onComplete = jest.fn();
      const close = jest.fn();

      let editorState = EditorState.createWithContent(convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: 'a',
            text: 'test',
          }
        ]
      }));
      let selection = editorState.getSelection();
      selection = selection.merge({
        focusOffset: 4,
      });
      editorState = EditorState.acceptSelection(editorState, selection);
      const wrapper = shallow((
        <ModalWorkflowSource
          editorState={editorState}
          entityType={{
            block: () => {},
          }}
          entity={{}}
          onComplete={onComplete}
          onClose={() => {}}
        />
      ));

      wrapper.instance().workflow = { close };
      wrapper.instance().onChosen({});

      expect(onComplete).toHaveBeenCalled();
      expect(AtomicBlockUtils.insertAtomicBlock).toHaveBeenCalled();
      expect(close).toHaveBeenCalled();

      AtomicBlockUtils.insertAtomicBlock.mockRestore();
    });

    it('prefer_this_title_as_link_text', () => {
      jest.spyOn(Modifier, 'replaceText');

      const onComplete = jest.fn();
      const close = jest.fn();

      let editorState = EditorState.createWithContent(convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: 'a',
            text: 'test',
          }
        ]
      }));
      let selection = editorState.getSelection();
      selection = selection.merge({
        focusOffset: 4,
      });
      editorState = EditorState.acceptSelection(editorState, selection);
      const wrapper = shallow((
        <ModalWorkflowSource
          editorState={editorState}
          entityType={{}}
          onComplete={onComplete}
          onClose={() => {}}
        />
      ));

      wrapper.instance().workflow = { close };
      wrapper.instance().onChosen({
        url: 'example.com',
        prefer_this_title_as_link_text: true,
      });

      expect(onComplete).toHaveBeenCalled();
      expect(Modifier.replaceText).toHaveBeenCalled();
      expect(close).toHaveBeenCalled();

      Modifier.replaceText.mockRestore();
    });
  });

  it('#onClose', () => {
    const onClose = jest.fn();
    const wrapper = shallow((
      <ModalWorkflowSource
        editorState={EditorState.createEmpty()}
        entityType={{}}
        entity={{}}
        onComplete={() => {}}
        onClose={onClose}
      />
    ));

    wrapper.instance().onClose({
      preventDefault: () => {},
    });

    expect(onClose).toHaveBeenCalled();
  });
});
