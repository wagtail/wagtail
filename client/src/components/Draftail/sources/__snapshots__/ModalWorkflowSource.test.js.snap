// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWorkflowSource #filterEntityData DOCUMENT 1`] = `
Object {
  "filename": "test.pdf",
  "id": 1,
  "url": "/documents/1/test.pdf",
}
`;

exports[`ModalWorkflowSource #filterEntityData EMBED 1`] = `
Object {
  "authorName": "Test",
  "embedType": "video",
  "providerName": "YouTube",
  "thumbnail": "https://i.ytimg.com/vi/pSlVtxLOYiM/hqdefault.jpg",
  "title": "Test",
  "url": "https://www.youtube.com/watch?v=pSlVtxLOYiM",
}
`;

exports[`ModalWorkflowSource #filterEntityData IMAGE 1`] = `
Object {
  "alt": "Test",
  "format": "right",
  "id": 53,
  "src": "/media/images/test.width-500.jpg",
}
`;

exports[`ModalWorkflowSource #filterEntityData LINK external 1`] = `
Object {
  "url": "https://www.example.com/",
}
`;

exports[`ModalWorkflowSource #filterEntityData LINK mail 1`] = `
Object {
  "url": "mailto:test@example.com",
}
`;

exports[`ModalWorkflowSource #filterEntityData LINK page 1`] = `
Object {
  "id": 60,
  "parentId": 1,
  "url": "/",
}
`;

exports[`ModalWorkflowSource #getChooserConfig LINK external 1`] = `
Object {
  "onload": Object {
    "type": "page",
  },
  "url": "/admin/choose-external-link/",
  "urlParams": Object {
    "allow_email_link": true,
    "allow_external_link": true,
    "can_choose_root": "false",
    "link_text": "",
    "link_url": "https://www.example.com/",
    "page_type": "wagtailcore.page",
  },
}
`;

exports[`ModalWorkflowSource #getChooserConfig LINK mail 1`] = `
Object {
  "onload": Object {
    "type": "page",
  },
  "url": "/admin/choose-email-link/",
  "urlParams": Object {
    "allow_email_link": true,
    "allow_external_link": true,
    "can_choose_root": "false",
    "link_text": "",
    "link_url": "test@example.com",
    "page_type": "wagtailcore.page",
  },
}
`;

exports[`ModalWorkflowSource #getChooserConfig LINK no entity 1`] = `
Object {
  "onload": Object {
    "type": "page",
  },
  "url": "/admin/choose-page/",
  "urlParams": Object {
    "allow_email_link": true,
    "allow_external_link": true,
    "can_choose_root": "false",
    "link_text": "",
    "page_type": "wagtailcore.page",
  },
}
`;

exports[`ModalWorkflowSource #getChooserConfig LINK page 1`] = `
Object {
  "onload": Object {
    "type": "page",
  },
  "url": "/admin/choose-page/0/",
  "urlParams": Object {
    "allow_email_link": true,
    "allow_external_link": true,
    "can_choose_root": "false",
    "link_text": "",
    "page_type": "wagtailcore.page",
  },
}
`;

exports[`ModalWorkflowSource works 1`] = `""`;
